crates/subspace-networking/src/utils.rs

```diff
@@ -95,3 +97,148 @@ pub(crate) fn convert_multiaddresses(addresses: Vec<Multiaddr>) -> Vec<PeerAddre
         })
         .collect()
 }
+
+/// Semaphore like implementation that allows both shrinking and expanding
+/// the max permits.
+#[derive(Clone, Debug)]
+pub(crate) struct ResizableSemaphore(Arc<SemShared>);
+
+/// The permit.
+#[derive(Clone, Debug)]
+pub(crate) struct ResizableSemaphorePermit(Arc<SemShared>);
+
+/// The state shared between the semaphore and the outstanding permits.
+#[derive(Debug)]
+struct SemShared {
+    /// The tuple holds (current usage, current max capacity)
+    state: Mutex<SemState>,
+
+    /// To signal waiters for permits to be available
+    notify: Notify,
+}
+
+/// The semaphore state.
+#[derive(Debug)]
+struct SemState {
+    /// The current capacity
+    capacity: usize,
+
+    /// The current outstanding permits
+    usage: usize,
+}
+
+impl SemState {
+    // Allocates a permit if available.
+    // Returns true if allocated, false otherwise.
+    fn alloc_one(&mut self) -> bool {
+        if self.usage < self.capacity {
+            self.usage += 1;
+            true
+        } else {
+            false
+        }
+    }
+
+    // Returns a free permit to the free pool.
+    // Returns true if any waiters need to be notified.
+    fn free_one(&mut self) -> bool {
+        let prev_is_full = self.is_full();
+        if let Some(dec) = self.usage.checked_sub(1) {
+            self.usage = dec;
+        } else {
+            panic!("SemState::free_one(): invalid free, state = {:?}", self);
+        }
+
+        // Notify if we did a full -> available transition.
+        prev_is_full && !self.is_full()
+    }
+
+    // Expands the max capacity by delta.
+    // Returns true if any waiters need to be notified.
+    fn expand(&mut self, delta: usize) -> bool {
+        let prev_is_full = self.is_full();
+        self.capacity += delta;
+
+        // Notify if we did a full -> available transition.
+        prev_is_full && !self.is_full()
+    }
+
+    // Shrinks the max capacity by delta.
+    fn shrink(&mut self, delta: usize) {
+        if let Some(dec) = self.capacity.checked_sub(delta) {
+            self.capacity = dec;
+        } else {
+            panic!("SemState::shrink(): invalid shrink, state = {:?}", self);
+        }
+    }
+
+    // Returns true if current usage exceeds capacity
+    fn is_full(&self) -> bool {
+        self.usage >= self.capacity
+    }
+}
+
+impl ResizableSemaphore {
+    pub(crate) fn new(capacity: usize) -> Self {
+        let shared = SemShared {
+            state: Mutex::new(SemState { capacity, usage: 0 }),
+            notify: Notify::new(),
+        };
+        Self(Arc::new(shared))
+    }
+
+    // Acquires a permit. Waits until a permit is available.
+    pub(crate) async fn acquire(&self) -> ResizableSemaphorePermit {
+        loop {
+            let wait = {
+                let mut state = self.0.state.lock().unwrap();
```

Comment:
We don't use `.unwrap()` except in tests and examples. We use `.expect()` and provide a proof in the message or explanation why this will never happen or in rare cases why it is okay to panic here.

Though here use of `parking_lot::Mutex` (that we prefer everywhere) would remove the need to unwrap in the first place.

Reply:
Looks like parking_lot doesn't have the issue of poisoned locks, which is nice

Reply:
The issue as such certainly exists, but `parking_lot` made a design decision different from std to not handle it explicitly.

Reply:
yeah they seem to drop the lock on thread panic, which keeps the callers clean.